uobtheatre/bookings/forms.py

```diff
@@ -0,0 +1,140 @@
+    """Input for creating Tickets with mutations."""
+
+    seat_group_id = IdInputField(required=True)
+    concession_type_id = IdInputField(required=True)
+    seat_id = IdInputField(required=False)
+    id = IdInputField(required=False)
+
+    def to_ticket(self):
+        """Returns the Ticket object to be updated/created.
+
+        If a Ticket already exists it is the exisitng Ticket object (saved in
+        the database). If not a new Ticket object is created (and not yet saved
+        to the database).
+
+        Returns:
+            Ticket: The ticket to be updated/created.
+        """
+
+        if self.id is not None:
+            return Ticket.objects.get(id=self.id)
+        return Ticket(
+            seat_group=SeatGroup.objects.get(id=self.seat_group_id),
+            concession_type=ConcessionType.objects.get(id=self.concession_type_id),
+            seat=Seat.objects.get(id=self.seat_id)
+            if self.seat_id is not None
+            else None,
+        )
+
+
+class TicketInputField(Field):
+    pass
+
+
+@convert_form_field.register(TicketInputField)
+def convert_form_field_to_string(field):
+    return graphene.List(
+        TicketInputType, description=field.help_text, required=field.required
+    )
+
+
+class BookingForm(MutationForm):
+    """Form for creating/updating a booking"""
+
+    tickets = TicketInputField(required=False)
+    user_email = CharField(required=False)
+
+    def clean(self):
+        """Validate form data on clean"""
+        cleaned_data = super().clean()
+        ## Various fields
+        if not self.instance.creator_id:
+            self.instance.creator = self.user
+
+        if cleaned_data.get("user_email"):
+            target_user, _ = User.objects.get_or_create(
+                email=cleaned_data.get("user_email"),
+                defaults={"first_name": "Anonymous", "last_name": "User"},
+            )
+        else:
+            target_user = self.user
+
+        if cleaned_data.get("user_email") or not self.instance.user_id:
+            self.instance.user = target_user
+
+        ## Check tickets
+        if cleaned_data.get("tickets") is not None:
+            ticket_objects = list(
+                map(lambda ticket: ticket.to_ticket(), cleaned_data.get("tickets"))
+            )
+
+            (
+                add_tickets,
+                delete_tickets,
+                total_number_of_tickets,
+            ) = self.instance.get_ticket_diff(ticket_objects)
+
+            self.cleaned_data["add_tickets"] = add_tickets
+            self.cleaned_data["delete_tickets"] = delete_tickets
+
+            if (
+                total_number_of_tickets > max_tickets_per_booking()
+                and not self.user.has_perm(
+                    "boxoffice", cleaned_data.get("performance").production
+                )
+            ):
+                raise ValidationError(
+                    {
+                        "tickets": "You may only book a maximum of %s tickets"
```

Comment:
1. Yes
2. Not sure - possibly so that the variable can't be mutated (accidentally or otherwise). Possibly to allow a change in the future to where this number is dependent on the user, but this would probably be done elsewhere. This hasn't changed in this PR.

Reply:
Feels like it might be worth defining this on the user model which takes in a performance and then returns an optional maximum?

Reply:
I kind of agree, but I'm not sure the user model is the best place. Typically I think something like this would live at the controller level, so I think I'll implement something in the form